cli/tools/registry/graph.rs

```diff
@@ -12,137 +16,189 @@ use deno_semver::jsr::JsrPackageReqReference;
 use deno_semver::npm::NpmPackageReqReference;
 use lsp_types::Url;
 
+use crate::cache::ParsedSourceCache;
+
 use super::diagnostics::PublishDiagnostic;
 use super::diagnostics::PublishDiagnosticsCollector;
 
-pub fn collect_invalid_external_imports(
-  graph: &ModuleGraph,
-  diagnostics_collector: &PublishDiagnosticsCollector,
-) {
-  let mut visited = HashSet::new();
-  let mut skip_specifiers: HashSet<Url> = HashSet::new();
-
-  let mut collect_if_invalid =
-    |skip_specifiers: &mut HashSet<Url>,
-     source_text: &Arc<str>,
-     specifier_text: &str,
-     resolution: &ResolutionResolved| {
-      if visited.insert(resolution.specifier.clone()) {
-        match resolution.specifier.scheme() {
-          "file" | "data" | "node" => {}
-          "jsr" => {
-            skip_specifiers.insert(resolution.specifier.clone());
-
-            // check for a missing version constraint
-            if let Ok(jsr_req_ref) =
-              JsrPackageReqReference::from_specifier(&resolution.specifier)
-            {
-              if jsr_req_ref.req().version_req.version_text() == "*" {
-                let maybe_version = graph
-                  .packages
-                  .mappings()
-                  .find(|(req, _)| *req == jsr_req_ref.req())
-                  .map(|(_, nv)| nv.version.clone());
-                diagnostics_collector.push(
-                  PublishDiagnostic::MissingConstraint {
-                    specifier: resolution.specifier.clone(),
-                    specifier_text: specifier_text.to_string(),
-                    resolved_version: maybe_version,
-                    text_info: SourceTextInfo::new(source_text.clone()),
-                    referrer: resolution.range.clone(),
-                  },
-                );
+pub struct GraphDiagnosticsCollector {
+  parsed_source_cache: Arc<ParsedSourceCache>,
```

Comment:
Got tired of passing dependencies down through all the functions (very viral) so changed this to use structs with deps that can be injected at a high level.